File: src/main/java/at/hrechny/predictionsbot/database/entity/MatchEntity.java
package at.hrechny.predictionsbot.database.entity;

import at.hrechny.predictionsbot.database.model.MatchStatus;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@EqualsAndHashCode(callSuper = true)
@Entity
@Table(name = "matches")
public class MatchEntity extends GeneratedIdEntity {

  @ManyToOne
  @JoinColumn(name="round_id", nullable=false)
  private RoundEntity round;

  @Column
  @Enumerated(EnumType.STRING)
  private MatchStatus status;

  @Column(columnDefinition = "TIMESTAMP")
  private Instant startTime;

  @ManyToOne
  @JoinColumn(name="home_team_id", nullable=false)
  private TeamEntity homeTeam;

  @ManyToOne
  @JoinColumn(name="away_team_id", nullable=false)
  private TeamEntity awayTeam;

  @Column
  private Integer homeTeamScore;

  @Column
  private Integer awayTeamScore;

  @Column(unique=true)
  private Long apiFootballId;

  @OneToMany(mappedBy="match", fetch = FetchType.EAGER, cascade = { CascadeType.PERSIST, CascadeType.MERGE })
  private List<PredictionEntity> predictions = new ArrayList<>();

  public Optional<PredictionEntity> getPrediction(Long userId) {
    return predictions.stream().filter(p -> p.getUser().getId().equals(userId)).findFirst();
  }

}

File: src/main/kotlin/at/hrechny/predictionsbot/database/entity/MatchEntity.kt
package at.hrechny.predictionsbot.database.entity

import at.hrechny.predictionsbot.database.model.MatchStatus
import io.micronaut.core.annotation.Introspected
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import java.time.Instant
import java.util.Optional

@Introspected(accessKind = [Introspected.AccessKind.FIELD])
@Entity
@Table(name = "matches")
class MatchEntity : GeneratedIdEntity() {
    @field:ManyToOne
    @field:JoinColumn(name = "round_id", nullable = false)
    var round: RoundEntity? = null

    @field:Column
    @field:Enumerated(EnumType.STRING)
    var status: MatchStatus? = null

    @field:Column(columnDefinition = "TIMESTAMP")
    var startTime: Instant? = null

    @field:ManyToOne
    @field:JoinColumn(name = "home_team_id", nullable = false)
    var homeTeam: TeamEntity? = null

    @field:ManyToOne
    @field:JoinColumn(name = "away_team_id", nullable = false)
    var awayTeam: TeamEntity? = null

    @field:Column
    var homeTeamScore: Int? = null

    @field:Column
    var awayTeamScore: Int? = null

    @field:Column(unique = true)
    var apiFootballId: Long? = null

    @field:OneToMany(mappedBy = "match", cascade = [CascadeType.PERSIST, CascadeType.MERGE])
    var predictions: MutableList<PredictionEntity> = mutableListOf()

    fun getPrediction(userId: Long): Optional<PredictionEntity> =
        predictions.stream().filter { prediction -> prediction.user?.id == userId }.findFirst()
}
